src/coordinator/storage/converter.go

```diff
@@ -177,3 +181,116 @@ func SeriesToPromSamples(series *ts.Series) []*prompb.Sample {
 
 	return samples
 }
+
+const (
+	// TODO(arnikola) get from config
+	initRawFetchAllocSize = 32
+)
+
+func iteratorToTsSeries(
+	iter encoding.SeriesIterator,
+	namespace ident.ID,
+) (*ts.Series, error) {
+	if namespace == nil {
+		namespace = iter.Namespace()
+	}
+	metric, err := FromM3IdentToMetric(namespace, iter.ID(), iter.Tags())
+	if err != nil {
+		return nil, err
+	}
+
+	datapoints := make(ts.Datapoints, 0, initRawFetchAllocSize)
+	for iter.Next() {
+		dp, _, _ := iter.Current()
+		datapoints = append(datapoints, ts.Datapoint{Timestamp: dp.Timestamp, Value: dp.Value})
+	}
+
+	return ts.NewSeries(metric.ID, datapoints, metric.Tags), nil
+}
+
+// Fall back to sequential decompression if unable to decompress concurrently
+func decompressSequentially(
+	iterLength int,
+	iters []encoding.SeriesIterator,
+	namespace ident.ID,
+) (*FetchResult, error) {
+	seriesList := make([]*ts.Series, iterLength)
+	for i, iter := range iters {
+		series, err := iteratorToTsSeries(iter, namespace)
+		if err != nil {
+			return nil, err
+		}
+		seriesList[i] = series
+	}
+
+	return &FetchResult{
+		SeriesList: seriesList,
+	}, nil
+}
+
+// SeriesIteratorsToFetchResult converts SeriesIterators into a fetch result
+func SeriesIteratorsToFetchResult(
+	seriesIterators encoding.SeriesIterators,
+	namespace ident.ID,
+	workerPools pool.ObjectPool,
+) (*FetchResult, error) {
+	defer seriesIterators.Close()
+
+	iters := seriesIterators.Iters()
+	iterLength := seriesIterators.Len()
+	seriesList := make([]*ts.Series, iterLength)
+
+	if workerPools == nil {
+		return decompressSequentially(iterLength, iters, namespace)
+	}
+	pool, ok := workerPools.Get().(xsync.WorkerPool)
+	if !ok {
+		return decompressSequentially(iterLength, iters, namespace)
+	}
+
+	var wg sync.WaitGroup
```

Comment:
can you extract this out into separate function and maybe call it decompressConcurrently ?